src/couch/src/couch_password_hasher.erl

```diff
@@ -21,20 +21,41 @@
     init/1,
     handle_call/3,
     handle_cast/2,
+    handle_info/2,
     code_change/3
 ]).
 
--export([hash/1]).
+-export([maybe_upgrade_password_hash/4, hash_admin_passwords/1]).
 
--record(state, {}).
+-export([worker_loop/1]).
+
+-define(IN_PROGRESS_ETS, couch_password_hasher_in_progress).
+
+-record(state, {
+    worker_pid,
+    worker_mon
+}).
 
 %%%===================================================================
 %%% Public functions
 %%%===================================================================
 
--spec hash(Persist :: boolean()) -> Reply :: term().
-hash(Persist) ->
-    gen_server:cast(?MODULE, {hash_passwords, Persist}).
+maybe_upgrade_password_hash(AuthModule, UserName, Password, UserProps) ->
+    IsAdmin = is_admin(UserProps),
+    NeedsUpgrade = needs_upgrade(UserProps),
+    InProgress = in_progress(AuthModule, UserName),
+    if
+        not IsAdmin andalso NeedsUpgrade andalso not InProgress ->
```

Comment:
We upgrade user doc passwords but not admins. Would it make sense to have a way to upgrade those too?

Reply:
That makes sense. We could put it in a comment in the code perhaps right above `maybe_upgrade_password_hash`, just so we if we change the code it wouldn't get out of sync with a readme.md file somewhere.

Reply:
I might try the admin upgrade to see what it looks like, now that I added the piece to preserve salt for upgrades I think I can mitigate any problems with the upgrade, it's just whether the fake admin doc makes the update harder, etc. but I would happily proceed with merging without it.